OCaml build support: add oasis build system

mgttlinger opened the pull request:
###### Motivation for this change

Some OCaml packages are built using oasis leading to a lot of repetition in the derivation code.

(I will create PRs for pgsolver and its dependencies once this change is merged)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mgttlinger committed:
OCaml build support: add oasis build system

vbgl commented:
May you please show a use-case for this `buildOasisPackage` function? Either as a separate commit in this PR or in a different (draft) PR based on this one.

mgttlinger committed:
tcslib: init at version v0.3

build for tcslib as contribution as well as usage example for the new buildOasisPackage

mgttlinger commented:
@vbgl as you can see in [the commit](https://github.com/NixOS/nixpkgs/pull/65715/commits/d940b4413794f85435ccb02be40d288faefedf36) this build helper effectively simplifies writing derivations for packages using oasis to specifying the source and dependencies without having to touch the build itself if the package doesn't do anything non standard.

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -0,0 +1,31 @@
+{ name, version, buildInputs ? [], ...
+}@args:
+
+buildOcaml (args // {
```
Please use `stdenv.mkDerivation` directly.

mgttlinger replied:
But then I would duplicate the code in `buildOcaml`. Is that intended? If so could you quickly tell me the reasoning behind that?

vbgl replied:
The `buildOcaml` helper function has been introduced for a specific set of libraries (the ones from Jane Street, at version approximately 113.33.xx). These legacy libraries are only available (in nixpkgs) for OCaml 4.02 and are bound to disappear.

Adding new packages that depend on this technical dept just creates a maintenance burden.

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -0,0 +1,31 @@
+{ buildOcaml, ocaml_oasis }:
+
+{ name, version, buildInputs ? [], ...
```
The variables `name` and `version` are bound but never used. Is it on purpose?

mgttlinger replied:
You're right. lets change it to:

```suggestion
{buildInputs ? [], ...
```

vbgl commented on pkgs/development/ocaml-modules/tcslib/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ stdenv, fetchFromGitHub, ocamlPackages, buildOasisPackage }:
```
`stdenv` is not required: `lib` is enough.

vbgl commented on pkgs/development/ocaml-modules/tcslib/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ stdenv, fetchFromGitHub, ocamlPackages, buildOasisPackage }:
```
Each OCaml library should be a different argument: `ounit`, `ocaml_extlib` and `num`; rather than the whole set `ocamlPackages`. Otherwise it will not be possible to build the package for other versions of OCaml.

mgttlinger replied:
Oh OK I thought that would be handled further up in the path e.g. `ocamlPackages` already containing the package set for the "correct" version

vbgl commented on pkgs/development/ocaml-modules/tcslib/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+buildOasisPackage rec {
+  name = "tcslib";
+  version = "v0.3";
```
This is not a valid version. It should start with a digit.

mgttlinger committed:
Suggested changes

mgttlinger committed:
Don't use buildOcaml

mgttlinger commented:
Curiously it seems like my last commit didn't get added to this PR. Any idea how to manually add that?

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -1,10 +1,24 @@
+          stdenv.lib.versionOlder minimumOcamlVersion ocaml.version;
+
+stdenv.mkDerivation (args // {
+  name = "ocaml-${name}-${version}";
```
Please include the version of OCaml in the name attribute.

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -1,10 +1,24 @@
+stdenv.mkDerivation (args // {
+  name = "ocaml-${name}-${version}";
+
+  buildInputs = [ ocaml findlib ocamlbuild camlp4 ocaml_oasis ] ++ buildInputs;
```
Are ocamlbuild and camlp4 needed?

mgttlinger replied:
ocamlbuild is needed

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -28,4 +42,11 @@ buildOcaml (args // {
     runHook postInstall
   '';
+
+  setupHook = if setupHook == null && hasSharedObjects
```
What’s the purpose of this hook?

mgttlinger replied:
Not sure to be honest. Added it because `buildOcaml` had it and I assumed the folks that wrote that had their reasons for having it. But I suppose if it turns out that it is needed (not in my uses so far) people can just send a PR and add it.

mgttlinger committed:
Suggested edits

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -0,0 +1,44 @@
+{ stdenv, ocaml_oasis, ocaml, findlib, ocamlbuild }:
+
+{ name, version, buildInputs ? [], meta ? { platforms = ocaml.meta.platforms or []; },
+  minimumOcamlVersion ? null,
```
OCaml is spelled with a capital ‘C’.

vbgl commented on pkgs/build-support/ocaml/oasis.nix:
```diff
@@ -0,0 +1,44 @@
+}@args:
+
+ assert minimumOcamlVersion != null ->
+          stdenv.lib.versionOlder minimumOcamlVersion ocaml.version;
```
Please provide a more explicit error message in case of assertion failure.

vbgl commented on pkgs/development/ocaml-modules/tcslib/default.nix:
```diff
@@ -0,0 +1,25 @@
+{ lib, fetchFromGitHub, ocamlPackages, buildOasisPackage, ounit, ocaml_extlib, num }:
+
+buildOasisPackage rec {
+  name = "tcslib";
```
Common practice is to call this attribute “pname”, as it lacks the version information.

mgttlinger committed:
Suggested edits

mgttlinger commented:
@vbgl Thank you for all the improvements and sorry that it takes so many iterations to get things right.

vbgl commented:
@GrahamcOfBorg build ocamlPackages.tcslib

vbgl commented:
Many thanks for your contribution. Good luck for pgsolver!
